File: Problems/Algorithms/139. Word Break/word_break.rs
impl Solution {
    pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
        let n = s.len();
        let mut dp = vec![false; n+1];
        dp[0] = true;
        
        for i in 1..=n {
            for j in 0..i {
                if dp[j] && word_dict.contains(&s[j..i].to_owned()) {
                    dp[i] = true;
                    break;
                }
            }
        }
        
        dp[n]
    }
}

File: Problems/Algorithms/139. Word Break/word-break.cpp
#include <string>
#include <vector>
#include <queue>
#include <unordered_set>

using namespace std;

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.length();

        unordered_set<string> words(wordDict.begin(), wordDict.end());
        vector<bool> visited(n, false);

        queue<int> queue;
        queue.push(0);

        while (!queue.empty()) {
            int start = queue.front();
            queue.pop();

            if (start == n) {
                return true;
            }

            for (int end = start + 1; end <= n; end++) {
                if (visited[end]) {
                    continue;
                }

                if (words.find(s.substr(start, end - start)) != words.end()) {
                    queue.push(end);
                    visited[end] = true;
                }
            }
        }

        return false;
    }
};
